### PR TITLE
Fix copy invite room message

### PR DESCRIPTION
**Pull request by dukenv0307**
### Details

Fix copy invite room message
### Fixed Issues

$ https://github.com/Expensify/App/issues/30272
PROPOSAL: https://github.com/Expensify/App/issues/30272#issuecomment-1777621410

### Tests

1. Create a room
2. Click on the header of the room > Invite
3. Invite some users
4. After the invite message appears, hover over this message and click on copy to clipboard
5. Paste it to composer, verify that the message that is copied is the same as the invite message displayed
6. Go to admin room of the workspace
7. Copy policy log of invite action above
8. Paste it to composer, and verify that the message that is copied is the same as the invite message displayed
- [x] Verify that no errors appear in the JS console

### Offline tests

None
### QA Steps

1. Create a room
2. Click on the header of the room > Invite
3. Invite some users
4. After the invite message appears, hover over this message and click on copy to clipboard
5. Paste it to composer, verify that the message that is copied is the same as the invite message displayed
6. Go to admin room of the workspace
7. Copy policy log of invite action above
8. Paste it to composer, and verify that the message that is copied is the same as the invite message displayed
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/129500732/162b49ee-bf34-4df8-a0e7-58e8e773d50e

https://github.com/Expensify/App/assets/129500732/112a7867-6d3e-477b-b8f8-52987e86bd7d

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/129500732/b5f26c88-6b3e-4ae0-b2f7-1ed98cb52ee3

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/129500732/f123a672-ec97-4375-baa8-544fca58100e

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/129500732/8bb9ae93-477d-4009-97f8-0ec446c296b3

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/129500732/27763874-9904-4a7e-9ab2-cb373d061e65

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/129500732/e63f5a64-7a6d-477a-ab12-109ffffdcad7

</details>

### CODE REVIEW

File: src/libs/ReportUtils.js [hunk 1]
```diff
@@ -4043,6 +4044,21 @@ function getTaskAssigneeChatOnyxData(accountID, assigneeAccountID, taskReportID,
...
+ * @returns {String}
+ */
+function getMentionMessage(accountIDs) {
+    const listMention = _.map(accountIDs, (accountID) => {
+        const personalDetail = lodashGet(allPersonalDetails, accountID);
+        const displayNameOrLogin =
+            LocalePhoneNumber.formatPhoneNumber(lodashGet(personalDetail, 'login', '')) || lodashGet(personalDetail, 'displayName', '') || Localize.translateLocal('common.hidden');
+        return `@${displayNameOrLogin}`;
+    });
+    return listMention.join(' and ');
```
Comment (cubuspl42) [inline src/libs/ReportUtils.js:4059]: I think we should stick to the format with commas:

image

File: src/libs/ReportUtils.js [hunk 2]
```diff
@@ -4114,6 +4130,16 @@ function getIOUReportActionDisplayMessage(reportAction) {
     return displayMessage;
 }
 
+/**
+ * Return room channel log display message
+ * @param {Object} reportAction
+ * @returns {String}
+ */
+function getRoomChannelLogMemberMessage(reportAction) {
+    const title = reportAction.actionName === CONST.REPORT.ACTIONS.TYPE.ROOMCHANGELOG.INVITE_TO_ROOM ? 'invited' : 'removed';
```
Comment (cubuspl42) [inline src/libs/ReportUtils.js:4139]: `title` is an unfortunate name; maybe `prefix` or `actionPerformed`?

File: src/styles/fontFamily/multiFontFamily.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
+import getOperatingSystem from '@libs/getOperatingSystem';
```
Comment (cubuspl42) [inline src/styles/fontFamily/multiFontFamily.ts:1]: Why are we changing an unrelated file?

File: src/libs/ReportUtils.js [hunk 4]
```diff
@@ -4043,6 +4044,21 @@ function getTaskAssigneeChatOnyxData(accountID, assigneeAccountID, taskReportID,
     };
 }
 
+/**
+ * Return the mention message of a list accounntID
+ * @param {Array<String>} accountIDs
+ * @returns {String}
+ */
+function getMentionMessage(accountIDs) {
```
Comment (cubuspl42) [inline src/libs/ReportUtils.js:4052]: It's not really a message though... I think that this logic is an implementation detail of `getRoomChannelLogMemberMessage`. I'm not against extracting functions at all, it just seems more appropriate for this to be a local function _ inside_ `getRoomChannelLogMemberMessage`.

Having in my [the other thread](https://github.com/Expensify/App/pull/30898/files#r1383204202), I think we'll just have to slightly restructure the logic of building our string.

File: src/libs/ReportUtils.js [hunk 5]
```diff
@@ -4115,6 +4116,30 @@ function getIOUReportActionDisplayMessage(reportAction) {
     return displayMessage;
 }
 
+/**
+ * Return room channel log display message
+ * @param {Object} reportAction
+ * @returns {String}
+ */
+function getRoomChannelLogMemberMessage(reportAction) {
```
Comment (cubuspl42) [inline src/libs/ReportUtils.js:4124]: Looks better! But I have a suggestion to increase the simplicity of the function. Let me know if you agree.

```js
/**
 * Return room channel log display message
 * 
 * @param {Object} reportAction
 * @returns {String}
 */
function getRoomChannelLogMemberMessage(reportAction) {
    const actionPerformed = reportAction.actionName === CONST.REPORT.ACTIONS.TYPE.ROOMCHANGELOG.INVITE_TO_ROOM ? 'invited' : 'removed';

    const mentions = _.map(reportAction.originalMessage.targetAccountIDs, (accountID) => {
        const personalDetail = lodashGet(allPersonalDetails, accountID);
        const displayNameOrLogin =
            LocalePhoneNumber.formatPhoneNumber(lodashGet(personalDetail, 'login', '')) || lodashGet(personalDetail, 'displayName', '') || Localize.translateLocal('common.hidden');
        return `@${displayNameOrLogin}`;
    });

    const lastMention = mentions.pop();

    if (mentions.length === 0) {
        return `${actionPerformed} ${lastMention}`;
    }

    return `${actionPerformed} ${mentions.join(', ')} and ${lastMention}`;
}
```

File: src/libs/ReportUtils.js [hunk 6]
```diff
@@ -4118,26 +4118,31 @@ function getIOUReportActionDisplayMessage(reportAction) {
...
+
+    const lastMention = mentions.pop();
+
+    if (mentions.length === 0) {
+        return `${actionPerformed} ${lastMention}`;
     }
-    if (listMention.length === 1) {
-        lastPrefix = ' and ';
+
+    if (mentions.length === 1) {
```
[the remark this answers is not in the document]
Comment (cubuspl42) [reply src/libs/ReportUtils.js:4141]: Oh... I think I get it

[the remark this answers is not in the document]
Comment (cubuspl42) [reply src/libs/ReportUtils.js:4141]: Still, I think this logic is still slightly more readable (even if the advantage is smaller than I originally assumed)

**Comment by cubuspl42** [comment]
Bad news: I think we'll need to expand the scope of the PR to handle another variant of the `invited @foo, @bar, @baz` message, which is nearly the same as ours, but not quite.

In the `#admins` channel, we have a message variant like this:

image

**Comment by cubuspl42** [comment]
@dukenv0307 Great! We'll need to expand the testing steps. When you do this, please re-upload at least the Web platform video.

**Comment by cubuspl42** [comment]
I'm having trouble observing the invite logs in the `#admins` channel...

In one case, I got information about one user being invited to a room when I invited three users at once:
image

In another case, I got no update about adding inviting to a room.

Do you always get reliable logs in `#admins` about the invited users, no matter how many you add at once?

**Comment by cubuspl42** [comment]
> @cubuspl42 Can you share a video for this case?

https://github.com/Expensify/App/assets/2590174/1a62275b-5348-49be-9159-05206d48c6de

> invited @cubuspl42.1+1@gmail.com, @cubuspl42.1+3@gmail.com, and @cubuspl42.1+5@gmail.com

...but in `#admins` only:

> invited @cubuspl42.1+5@gmail.com to #social12

I'm not saying this is something your PR introduced, but it complicates the testing 🫤

**Comment by cubuspl42** [comment]
@dukenv0307 I'll re-test

**Comment by cubuspl42** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/2590174/19b7fa7b-3052-4147-8d45-9a13ffebeaa4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/2590174/631f033c-5037-4193-b968-b2973e930a7b

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/2590174/e98ed263-a41a-444a-b2bc-e5cdfb1be686

</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/2590174/fb1b83b1-0bde-4c54-9d67-adda0fef91f4

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/2590174/74d4bda6-6421-4bec-85c9-2d45acf0dd7c

</details>

**Review by jakub-trzebiatowski — APPROVED** [review]

**Comment by cubuspl42** [comment]
@dukenv0307 Thank you for reminding me. I submitted the checklist without approving.

**Review by bondydaa — APPROVED** [review]

File: src/libs/ReportUtils.js [hunk 7]
```diff
@@ -4130,6 +4131,48 @@ function getIOUReportActionDisplayMessage(reportAction) {
...
+        message = `${verb} ${mentions.join(', ')}, and ${lastMention}`;
+    }
+
+    const roomName = lodashGet(reportAction, 'originalMessage.roomName', '');
+    if (roomName) {
+        const preposition =
+            reportAction.actionName === CONST.REPORT.ACTIONS.TYPE.ROOMCHANGELOG.INVITE_TO_ROOM || reportAction.actionName === CONST.REPORT.ACTIONS.TYPE.POLICYCHANGELOG.INVITE_TO_ROOM
+                ? ' to'
+                : ' from';
+        message += `${preposition} ${roomName}`;
```
Comment (shubham1206agra) [inline src/libs/ReportUtils.js:4170]: Minor bug from here, the room name was not shown as link. More details here https://github.com/Expensify/App/issues/31568
